flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java

```diff
@@ -296,42 +292,31 @@ public void resetNFAChanged() {
 			if (shouldDiscardPath) {
 				// a stop state was reached in this branch. release branch which results in removing previous event from
 				// the buffer
-				for (final ComputationState<T> state : statesToRetain) {
-					eventSharedBuffer.release(
-							NFAStateNameHandler.getOriginalNameFromInternal(
-									state.getPreviousState().getName()),
-							state.getEvent(),
-							state.getTimestamp(),
-							state.getCounter());
+				for (final ComputationState state : statesToRetain) {
+					sharedBuffer.releaseNode(state.getPreviousBufferEntry());
 				}
 			} else {
 				computationStates.addAll(statesToRetain);
 			}
 
 		}
 
-		discardComputationStatesAccordingToStrategy(computationStates, result, afterMatchSkipStrategy);
-
-		// prune shared buffer based on window length
-		if (windowTime > 0L) {
-			long pruningTimestamp = timestamp - windowTime;
-
-			if (pruningTimestamp < timestamp) {
-				// the check is to guard against underflows
+		discardComputationStatesAccordingToStrategy(
+			sharedBuffer, computationStates, result, afterMatchSkipStrategy);
 
-				// remove all elements which are expired
-				// with respect to the window length
-				if (eventSharedBuffer.prune(pruningTimestamp)) {
-					nfaChanged = true;
-				}
-			}
+		if (event.getEvent() == null) {
```

Comment:
Why here `==null`? I may be missing sth.

Reply:
This is unfortunate way we currently push Watermark into NFA.

Reply:
I see. Then at least we should have a comment there explaining it.

Reply:
actually I would recommend having a method: `nfa.advanceTime()` or `nfa.clearUpTo(timestamp)` and we call it from the `AbstractKeyedCEPPatternOperator.advanceTime()` after calling `processEvent()`. This way it is clearer what it does and when it is called. What do you think?

Reply:
Agree, will add as a separate commit.